src/advanced/domains/coupon/components/CouponForm.tsx

```diff
@@ -0,0 +1,137 @@
+import type { ChangeEvent, FocusEvent, FormEvent } from "react";
+
+import { Button, SearchInput, useNotifications } from "../../../shared";
+
+type CouponFormType = {
+  name: string;
+  code: string;
+  discountType: "amount" | "percentage";
+  discountValue: number;
+};
+
+type CouponFormProps = {
+  couponForm: CouponFormType;
+  onSubmit: (e: FormEvent) => void;
+  onCancel: () => void;
+  onFormChange: (form: CouponFormType) => void;
+};
+
+export function CouponForm({ couponForm, onSubmit, onCancel, onFormChange }: CouponFormProps) {
+  const { addNotification } = useNotifications();
+  const handleNameChange = (e: ChangeEvent<HTMLInputElement>) => {
+    onFormChange({
+      ...couponForm,
+      name: e.target.value
+    });
+  };
+
+  const handleCodeChange = (e: ChangeEvent<HTMLInputElement>) => {
+    onFormChange({
+      ...couponForm,
+      code: e.target.value.toUpperCase()
+    });
+  };
+
+  const handleDiscountTypeChange = (e: ChangeEvent<HTMLSelectElement>) => {
+    onFormChange({
+      ...couponForm,
+      discountType: e.target.value as "amount" | "percentage"
+    });
+  };
+
+  const handleDiscountValueChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const value = e.target.value;
+
+    if (value === "" || /^\d+$/.test(value)) {
```

Comment:
요렇게 가독성을 좀 더 높여봐도 좋을 것 같습니다.

```suggestion
    const isNumbericString = /^\d+$/.test;
    if (value === "" || isNumberString(value)) {
```